### DISCUSSION TITLE
should on_message cloud function be a child or a sibling of queue

### DISCUSSION THREAD

**Discussion by ekeren** [winglang/wing#780] [General]
We have the following code in our examples:

```typescript
bring cloud;

let queue = new cloud.Queue();
let bucket = new cloud.Bucket();

queue.on_message((m: str): str ~> {
  bucket.put(“wing.txt”, “Hello ${m}!”)
});

```

The structure that it currently produces in the console is:

```sh
Root
|- Bucket
|- Queue
    |- onMessage-62363464
```

Meaning that the cloud function that saves the file on the bucket is a child of the queue.
I think that it is more correct that a queue is just a leaf, and doesn’t own the cloud function, I believe that the proper representation of this hierarchy should be:
```sh
Root
|- Queue
|- Bucket
|- Function 
```

Which means that the on_message function should not accept an `inflight function`, it should accept a `cloud.Function`
and the queue `on_message` property is set to that `cloud.Function` but the queue itself is not the functions' parent

A more descriptive way to write the above code will be:

```typescript
bring cloud

let queue = new cloud.Queue();
let bucket = new cloud.Bucket() be store-greeting;
let handler = new cloud.Function((m: str): str ~> {
  bucket.put(“${m}.txt”, “Hello ${m}!”)
}) be "execute-greetings";

queue.on_message(handler);
```
And the corresponding hierarchy will be:

```sh
Root
|- Queue
|- store-greeting
|- execute-greetings 
```

The queue itself is just an entry point to this greeting system, it should not contain the actual logic.

**Comment by staycoolcall911** [comment]
@ekeren - the last line of the 2nd code sample should be `queue.on_message(handler);`, right?

**Reply by ekeren** [reply]
sure, updated. Thanks

**Comment by Chriscbr** [comment]
If I understand correctly, there's two issues being discussed - one is whether any function resource added/attached by `on_message` should be a sibling or child of the queue, and the other is whether the method should accept an inflight or a function resource.

The first issue is pretty straight forward - I totally agree, currently the way the resource is generated as a child of the queue doesn't totally make sense. Luckily this is a pretty easy fix. I'll write up a PR to update this.

The second issue is a bit more tricky. In an early iteration of the Queue resource, we actually did have the `on_message` method accept a `cloud.Function` resource like you described. But I found some technical issues with this model and compiling it to the cloud. I'm happy to give more details, but it will probably take a bit more time to write up.

**TL;DR** - subscribing to the events of a queue or other resource usually means there needs to exist code somewhere to convert from a cloud provider's native event structures (like [SQSEvent](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/65043c2b84d021b0d671661f6f1dd693933d1d47/types/aws-lambda/trigger/sqs.d.ts#L6-L22)) to the Wing SDK types. The most straightforward way to do this is to add a wrapper to the user's code to preprocess inputs, but this would require modifying user's code, and **changing** the effective type of the user's resource from `cloud.Function<input: str>` to `cloud.Function<input: SQSEvent>`. To avoid this surprising change, the current API instead expects an inflight with `input: str` as a parameter, and generates the `cloud.Function<input: SQSEvent>` for you. There are other ways we could implement this without adding wrappers to user code (for example, performing data transformations outside of user code, like in an additional lambda function), but these solutions might have performance costs.

**Reply by Chriscbr** [reply]
BTW, if enough users really want it, we _could_ offer both APIs, just with different implementations. For example:

- `queue.on_message(inflight)` (most performant, recommended, 1 FaaS is generated by compiler)
- `queue.on_message_fn(fn)` (less performant, desugars into `let temp = (event) ~> { fn.invoke(event) }; queue.on_message(temp)`, 2 FaaS are generated by compiler)

**Reply by ekeren** [reply]
It is interesting that you describe it as two issues, after writing this I started thinking that the only issues in question is wether the queue accepts an `inflight function` or a `cloud.Function`.

The way I see it, this it what determines whether it is a sibling or a parent relationship.

I understands the technical difficulty you describe but I want first to answer the, somewhat puristic, question of responsibility, who is responsible to create the `cloud.Function` object? My gut instinct tells me it should not be the queue,  but it might be that `inflight function` is the type that describes computation that we pass along from one object to another. 

This is the example I have in mind to describe the difficulty of this question: 
```js
let compute_something = (s:str) : str ~> { //something of computing } 
let queue = new cloud.Queue();
let api = new cloud.ApiGateway();

queue.on_message(
  //TODO pass compute_something
);
api.on_post("/tirgger-computation",
   //TODO pass compute_something
) ;
```

When thinking about the TODO inside we need to cover some questions:
* Types: 
   queue.on_message probably needs a function from type `(e:cloud.QueueEvent) ~> void`
   api.on_post probably needs a function from type `(req: cloud.ApiGatewayPostRequest) ~> cloud.ApiGatewayPostRequest)`
   
* What is the best way to compute_something? (e.g. lambda function/ a fargate container) and does it make sense to determine this just once or twice

**Reply by Chriscbr** [reply]
yeah - from a puristic standpoint, I'm not sure I have a strong opinion on where the responsibility should be. In my mind I read `queue.on_message(...)` as just a declarative statement, saying -- hey, this is a computation that be run with messages from the queue. It's kind of like a declarative version of `queue.for_each(...)` (for_each is something I expect to happen "now").

On the other hand, I guess I do have a concern that implicitly creating a FaaS may be less ergonomic at best, or surprising at worst. What if I need to perform a batch job (1hr or longer) for each queue message? In that case

```ts
let queue = new cloud.Queue();
let job = new cloud.Job((event: QueueEvent) ~> { /* long computation */ });
queue.on_message(job);
```

looks better to me than:

```ts
let queue = new cloud.Queue();
let job = new cloud.Job((event: QueueEvent) ~> { /* long computation */);
queue.on_message((event: QueueEvent) ~> { job.start(event) });
```

**Reply by ekeren** [reply]
Another example of the responsibility constraints is what happens if I need to I state that the runtime require nodejs:18.*

This kind of requirement should be a cloud.Function requirement and is the concern of the person passing the handler to on_message I would  not expect for the queue to fulfill this requirement

**Reply by Chriscbr** [reply]
That's a valid concern - another example of this is if someone wants to add environment variables to their cloud function. Right now we support this through an additional props argument, I believe:

```ts
queue.on_message(
  (event) ~> { print(event) },
  env: { DEBUG: "true" }
);
```

**Reply by eladb** [reply]
BTW @ekeren for pure wing code, users shouldn't be able to specify the node runtime on functions. The fact that this compiles to JS and runs in node is an technically implementation details

**Reply by ekeren** [reply]
One reason for having to rely on a specific runtime is the fact that you bring a package that requires a specific node version or higher

**Reply by Chriscbr** [reply]
I think the version of the node/python/etc. runtime you're using might blur into more of an operational concern

**Comment by eladb** [comment]
I think we need both options: one that you to "bring your own worker" (`queue.add_worker()`), and one that is a ergonomic sugar that reduces boilerplate (`queue.on_message()`).

> I am not sure about the APIs, we need to discuss in a wider context as part of the SDK RFC.

As for resource scoping for the implicit `cloud.Function` (the topic of this discussion), I can see how it might be friendlier to put the implicitly created function as a sibling of the queue instead of its child. I am okay with that if we feel this makes more sense for users. I love it that Wing Console surfaces these questions :-).

P.S. We should have a deeper discussion on our *event-driven API pattern* as part of a **Wing SDK RFC**. @Chriscbr I know that @staycoolcall911 and you discussed kicking off this RFC, and I think we are ready to write it, and that it will really help bring up all these topics with a holistic view of the entire surface area.

**Reply by ekeren** [reply]
Don't you feel that there is a correlation between the identity of the "creator" of the `cloud.Function` resource and the hierarchy (sibling vs parent)

It seems unintuitive to me that I can create a sibling resource

**Reply by eladb** [reply]
I think it's okay to think about this as an aesthetic/ergonomic aspect of how the tree "looks" and how intuitive it is to navigate it rather than a puristic concern of maintaining strict ownership/scoping relationship. As long as the `cloud.Function` ID is derived from the queue (e.g. `${queue.id}-on_message`), then it would be okay to make it a sibling instead of a child.

**Comment by eladb** [comment]
Let's fold this discussion into #858 ?
